Y24-220 Create v2 end point for tagsets

seenanair opened the pull request:
Closes https://github.com/sanger/sequencescape/issues/4238

#### Changes proposed in this pull request

- API v2 end point to query tag sets
- Fetaure flag support (`y24_220_enable_tag_set_api)`

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

harrietc52 commented on app/resources/api/v2/tag_set_resource.rb:
```diff
@@ -0,0 +1,61 @@
+      filter :visible, default: true, apply: ->(records, _value, _options) { records.visible }
+    end
+  end
+end
```
It this resource is immutable, as all the attributes and associations are read-only? If so, `immutable` can be added to the top of the  resource file, and the `routes.rb` could be updated so that any other request types are denied.

seenanair replied:
Done 👍

harrietc52 commented on app/resources/api/v2/tag_set_resource.rb:
```diff
@@ -0,0 +1,61 @@
+    #
+    # This resource represents a TagSet, which links together two related tag groups.
+    # It includes the following attributes and relationships:
+    #
```
Would it be possible to include the example GET requests, including the filter examples?

I’ve added example requests as part of updated docs [here](https://github.com/sanger/sequencescape/pull/4687). You can then use the GitHub action to rebuild the Yard Syntax and view these comments/ examples [here](https://sanger.github.io/sequencescape/Api/V2.html)

harrietc52 replied:
I think you might need to run this [Workflow](https://github.com/sanger/sequencescape/actions/workflows/generate_pages.yml), as described [here](https://github.com/sanger/sequencescape?tab=readme-ov-file#documentation), to update the SS Yard documentation on GitHub pages :)

harrietc52 commented on spec/requests/api/v2/tag_sets_spec.rb:
```diff
@@ -0,0 +1,145 @@
+      )
+    end
+  end
+end
```
Great tests! Thank you.

harrietc52 reviewed (approved):
Looks good! 

Just the comment about adding the example requests for the Yard docs. 

Whilst adding comments/ example requests for the other API v2 requests, I created a PostMan directory to persist mainly the POST requests, but also the GETs. If you would like to (or just to have a play) you can add these GET requests to the PostMan. Ref [here](https://github.com/sanger/sequencescape/pull/4687/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R483)

KatyTaylor commented on app/api/endpoints/tag_sets.rb:
```diff
@@ -0,0 +1,7 @@
+  model {}
+
+  instance {}
+end
```
Was this necessary for some reason? We shouldn't be adding new V1 endpoints - Stuart's project was to replace remaining V1 use with V2 API.

seenanair replied:
I was not sure about this and I just copied how it was for other end points. :D It sounds like it is for supporting V1. Shall I remove this?

KatyTaylor replied:
Yes pretty sure this is just for V1 so unnecessary.

KatyTaylor commented on app/controllers/api/v2/tag_sets_controller.rb:
```diff
@@ -0,0 +1,19 @@
+
+      def check_feature_flag
+        render json: { error: 'TagSets API is disabled' } unless Flipper.enabled?(:y24_220_enable_tag_set_api)
+      end
```
Think there should there be an addition to `feature_flags.yml` in this PR?

KatyTaylor replied:
Also, if you haven't already, would you mind creating a story and a pull request to remove the feature flag, so it's easy to remove when we decide to? There's a team feature flags guidance page on Confluence in case if you haven't seen it.

seenanair replied:
Done 👍 https://github.com/sanger/sequencescape/issues/4766

KatyTaylor replied:
Thank you 🙏 

I think the idea was to go even further, and create the (draft) pull request ready to merge in whenever appropriate - so that removing the feature flag is really effortless (obviously extra effort for you upfront, but hopefully less overall since it's fresh in your head).

Also, thinking about when to remove the flag - on the story you've written "once the prerequisite work (Y25-185) is completed". 

We can definitely switch the new feature 'on' once Y25-185 is completed, but I think we could let the flag hang around for a couple of months or so afterwards, just to ensure the page gets used and we haven't missed anything?

This was what I thought about the level of risk (copied from Slack):

> I think there is a small remaining risk, as we are making the functionality on that page more constrained, in that we're not including ALL the tag groups, and we're not allowing you to select any possible combination of tag 1 and tag 2.
So feature flagging could help in that we could resolve any prod issues immediately by flipping the flag.

Since you've implemented the flag, I feel we may as well keep it around for a little while in case we get any production issues due to the above?

seenanair replied:
Oh, okay. I agree, and all of your comments make complete sense. I will create a draft PR and chang the timing to remove the flag.

seenanair replied:
@KatyTaylor 
Following draft PRs created which enable the removal of 
SS - https://github.com/sanger/sequencescape/pull/4770
Limber - https://github.com/sanger/limber/pull/2275
Integration suite - https://gitlab.internal.sanger.ac.uk/psd/integration-suite/-/merge_requests/232

KatyTaylor reviewed (changes requested):
Looks mostly good, thank you! Just a couple of questions to look at.

KatyTaylor reviewed (approved):
Looks good, thank you, although a couple of tests are still failing

seenanair commented:
@harrietc52 @KatyTaylor I've addressed all your comments, and now all tests are passing. Please review if required.

seenanair commented:
> > @harrietc52 @KatyTaylor I've addressed all your comments, and now all tests are passing. Please review and let me know if any further changes are needed.
> 
> Looks great! Just the command to run to update the SS GitHub Pages for Yard documentation as mentioned above 😄

@harrietc52 Thanks for the review! 🙏 I ran the action [14081314706](https://github.com/sanger/sequencescape/actions/runs/14081314706) against branch `y24_220_create_v2_end_point_for_tagsets` , but the documentation hasn’t been updated with the Tagset API. Did I miss something?

harrietc52 commented:
> > > @harrietc52 @KatyTaylor I've addressed all your comments, and now all tests are passing. Please review and let me know if any further changes are needed.
> > 
> > 
> > Looks great! Just the command to run to update the SS GitHub Pages for Yard documentation as mentioned above 😄
> 
> @harrietc52 Thanks for the review! 🙏 I ran the action [14081314706](https://github.com/sanger/sequencescape/actions/runs/14081314706) against branch `y24_220_create_v2_end_point_for_tagsets` , but the documentation hasn’t been updated with the Tagset API but worked fine locally. Did I miss something?

Oh strange! That is odd. I’ll try give it a go too
